src/main/kotlin/org/jetbrains/kotlin/jupyter/protocol.kt

```diff
@@ -169,12 +173,50 @@ fun JupyterConnection.Socket.shellMessagesHandler(msg: Message, repl: ReplForJup
     }
 }
 
-class CapturingOutputStream(val stdout: PrintStream, val captureOutput: Boolean) : OutputStream() {
+class CapturingOutputStream(private val stdout: PrintStream,
+                            private val conf: OutputConfig,
+                            private val captureOutput: Boolean,
+                            val onCaptured: (String) -> Unit) : OutputStream() {
     val capturedOutput = ByteArrayOutputStream()
+    private var time = System.currentTimeMillis()
+    private var overallOutputSize = 0
+    private var newlineFound = false
+
+    private fun shouldSend(b: Int): Boolean {
+        val c = b.toChar()
+        newlineFound = newlineFound || c == '\n' || c == '\r'
+        if (newlineFound && capturedOutput.size() >= conf.captureNewlineBufferSize)
+            return true
+        if (capturedOutput.size() >= conf.captureBufferMaxSize)
+            return true
+
+        val currentTime = System.currentTimeMillis()
+        if (currentTime - time >= conf.captureBufferTimeLimitMs) {
+            time = currentTime
+            return true
+        }
+        return false
+    }
 
     override fun write(b: Int) {
+        ++overallOutputSize
         stdout.write(b)
-        if (captureOutput) capturedOutput.write(b)
+
+        if (captureOutput && overallOutputSize <= conf.cellOutputMaxSize) {
+            capturedOutput.write(b)
+            if (shouldSend(b)) {
+                flush()
+            }
```

Comment:
What happens if no more output is produced, but the cell is still running? Who will flush the accumulated buffer?

Reply:
Added timer.